src/inference/kernel_dsl.jl

```diff
@@ -107,19 +104,16 @@ function expand_kern_ex(ex)
     end
 
     ex = quote
-        function $(esc(f))(
-                $(esc(trace))::Trace, $(toplevel_args...);
-                check=false, observations=EmptyChoiceMap())
+        function $(f)($(trace)::Trace, $(toplevel_args...);
+                      check=false, observations=EmptyChoiceMap())
```

Comment:
We might want to replace `check` and `observations` with `gensym`-ed versions just to be safe, and make sure we use the `gensym`-ed versions throughout the rest of the expression --- otherwise, there's a (small) chance that the `check` and `observations` names we introduced will clash with the names of the arguments that the user writes when defining the kernel. Normally macro expansion would handle the gensyms for us, but if we are escaping everything, then `check` will remain as `check`.

Reply:
This makes me think the best thing to do here is actually `esc` as few things as possible! So remove the `esc` of the final `Expr`. The only thing we need to be escaping are the user-provided names of the kernels themselves, because that's how we'll refer to them afterwards. We don't have to escape the user-provided argument names, either within the function header or within the function body --- macro-expansion should automatically do the right thing for us by gen-syming or not gen-syming them appropriately.

Reply:
> I think this is mostly making the right fixes, but we have to be careful about how `esc`aping the final `Expr` affects name resolution for names and references we're injecting into the generated code.
> 
> For example, if you escape a block of code which contains a line like `Gen.reversal(...)`, `Gen.reversal` may not always end up pointing to the right function, depending on what `Gen` refers to in the context where the macro is called. To avoid this, we should either avoid escaping lines like that, or if easier, replace `Gen.reversal` with `$(GlobalRef(Gen, :reversal))`.

I think in this context, more things are going to go wrong if the user writes `import Gen as G` and then tries to use the kernel DSL -- in fact, in that case, we need to annotate every single type with `G` (because `import Gen as G` does not define `Trace`, for example, in the scope of the import).

Reply:
I think annotating everything with `GlobalRef` is not necessary. I think it's overly conservative -- e.g. a lot of the functionality of Gen assumes the user writes `using Gen` _somewhere_ in their development module.

Given that we make that assumption elsewhere in the codebase, I don't see any reason to fix it here.

Reply:
Re -- I implemented `gensym` versions of the keyword arguments, and made these consistent in both the forward and reversal kernel.

Reply:
If we don't `esc` the final returned expression, then we don't have to do the `GlobalRef` annotation, so that's fine!

We do use the `GlobalRef` approach in other parts of the code though, particularly in codegen for the static modeling language -- see Issue #272 and the corresponding fix PR #273.